File: src/main/java/com/d20charactersheet/framework/dac/dao/sql/SqlAbilityDao.java
package com.d20charactersheet.framework.dac.dao.sql;

import static com.d20charactersheet.framework.dac.abilitybuilder.ExtraFeatsAbilityBuilder.KEY_NUMBER_OF_EXTRA_FEATS;
import static com.d20charactersheet.framework.dac.abilitybuilder.ExtraSkillPointsAbilityBuilder.KEY_NUMBER_OF_SKILL_POINTS;
import static com.d20charactersheet.framework.dac.abilitybuilder.SpelllistAbilityBuilder.ANY_SCHOOL_ID;
import static com.d20charactersheet.framework.dac.abilitybuilder.SpelllistAbilityBuilder.KEY_CASTING_TYPE_ID;
import static com.d20charactersheet.framework.dac.abilitybuilder.SpelllistAbilityBuilder.KEY_KNOWN_SPELLS_TABLE_ID;
import static com.d20charactersheet.framework.dac.abilitybuilder.SpelllistAbilityBuilder.KEY_SCHOOL_ID;
import static com.d20charactersheet.framework.dac.abilitybuilder.SpelllistAbilityBuilder.KEY_SPELLLIST_ID;
import static com.d20charactersheet.framework.dac.abilitybuilder.SpelllistAbilityBuilder.KEY_SPELLS_PER_DAY_TABLE_ID;
import static com.d20charactersheet.framework.dac.abilitybuilder.SpelllistAbilityBuilder.KEY_SPELL_ATTRIBUTE_ID;
import static com.d20charactersheet.framework.dac.abilitybuilder.SpelllistAbilityBuilder.KEY_SPELL_SOURCE_ID;
import static com.d20charactersheet.framework.dac.dao.sql.TableAndColumnNames.COLUMN_ABILITY_ID;
import static com.d20charactersheet.framework.dac.dao.sql.TableAndColumnNames.COLUMN_ABILITY_TYPE_ID;
import static com.d20charactersheet.framework.dac.dao.sql.TableAndColumnNames.COLUMN_CLASSNAME;
import static com.d20charactersheet.framework.dac.dao.sql.TableAndColumnNames.COLUMN_DESCRIPTION;
import static com.d20charactersheet.framework.dac.dao.sql.TableAndColumnNames.COLUMN_ID;
import static com.d20charactersheet.framework.dac.dao.sql.TableAndColumnNames.COLUMN_NAME;
import static com.d20charactersheet.framework.dac.dao.sql.TableAndColumnNames.COLUMN_PROPERTY_KEY;
import static com.d20charactersheet.framework.dac.dao.sql.TableAndColumnNames.COLUMN_PROPERTY_VALUE;
import static com.d20charactersheet.framework.dac.dao.sql.TableAndColumnNames.SELECT;
import static com.d20charactersheet.framework.dac.dao.sql.TableAndColumnNames.SQL_GET_ABILITY_PROPERTIES;
import static com.d20charactersheet.framework.dac.dao.sql.TableAndColumnNames.SQL_GET_ALL_ABILITIES;
import static com.d20charactersheet.framework.dac.dao.sql.TableAndColumnNames.SQL_WHERE_ID;
import static com.d20charactersheet.framework.dac.dao.sql.TableAndColumnNames.TABLE_ABILITY;
import static com.d20charactersheet.framework.dac.dao.sql.TableAndColumnNames.TABLE_ABILITY_PROPERTY;

import java.sql.SQLException;
import java.util.ArrayList;
import java.util.EnumSet;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import com.d20charactersheet.framework.boc.model.Ability;
import com.d20charactersheet.framework.boc.model.ExtraFeatsAbility;
import com.d20charactersheet.framework.boc.model.ExtraSkillPointsAbility;
import com.d20charactersheet.framework.boc.model.KnownSpellsTable;
import com.d20charactersheet.framework.boc.model.School;
import com.d20charactersheet.framework.boc.model.Spelllist;
import com.d20charactersheet.framework.boc.model.SpelllistAbility;
import com.d20charactersheet.framework.boc.model.SpellsPerDayTable;
import com.d20charactersheet.framework.dac.abilitybuilder.AbilityBuilder;
import com.d20charactersheet.framework.dac.abilitybuilder.AbilityBuilderFactory;
import com.d20charactersheet.framework.dac.abilitybuilder.AbilityConfig;
import com.d20charactersheet.framework.dac.abilitybuilder.DefaultAbilityBuilder;
import com.d20charactersheet.framework.dac.abilitybuilder.ExtraFeatsAbilityBuilder;
import com.d20charactersheet.framework.dac.abilitybuilder.ExtraSkillPointsAbilityBuilder;
import com.d20charactersheet.framework.dac.abilitybuilder.SpelllistAbilityBuilder;
import com.d20charactersheet.framework.dac.dao.AbilityDao;
import com.d20charactersheet.framework.dac.dao.sql.rowmapper.AbilityConfigRowMapper;

/**
 * Provides access to the ability tables in the SQLite database.
 */
public class SqlAbilityDao implements AbilityDao {

    private static final String SQL_GET_ID = SELECT + "id FROM " + TABLE_ABILITY + " WHERE rowid = ?";

    private static final String SQL_WHERE_ABILITY_ID_AND_PROPERTY_KEY =
        COLUMN_ABILITY_ID + " = ? AND " + COLUMN_PROPERTY_KEY + " = ?";

    private final Database db;
    private final RowMapper abilityConfigRowMapper;

    /**
     * Creates SQLiteAbilityDao accessing the given database.
     *
     * @param db
     *     The database to access.
     */
    public SqlAbilityDao(final Database db) {
        this.db = db;
        abilityConfigRowMapper = new AbilityConfigRowMapper();
    }

    @Override
    public List<Ability> getAllAbilities(final List<Spelllist> allSpelllists, final List<KnownSpellsTable> allKnownSpellsTables,
        final List<SpellsPerDayTable> allSpellsPerDayTables) {
        final List<Ability> allAbilities = new ArrayList<>();
        final AbilityBuilderFactory abilityBuilderFactoy = createAbilityBuilderFactory(allSpelllists, allKnownSpellsTables,
                                                                                       allSpellsPerDayTables);
        QueryResult queryResult = null;
        try {
            queryResult = db.rawQuery(SQL_GET_ALL_ABILITIES, new String[0]);
            for (queryResult.moveToFirst(); !queryResult.isAfterLast(); queryResult.moveToNext()) {
                final AbilityConfig abilityConfig = (AbilityConfig) abilityConfigRowMapper.mapRow(queryResult.getDataRow());
                selectAbilityProperties(abilityConfig);
                final Ability ability = createAbility(abilityConfig, abilityBuilderFactoy);
                allAbilities.add(ability);
            }
        } catch (final SQLException sqlException) {
            //            Logger.error("Can't get all abilities", sqlException);
        } finally {
            if (queryResult != null) {
                queryResult.close();
            }
        }
        return allAbilities;
    }

    private void selectAbilityProperties(final AbilityConfig abilityConfig) {
        if (!abilityConfig.getClassname().equals(Ability.class.getSimpleName())) {
            final Map<String, String> abilityProperties = new HashMap<>();
            QueryResult queryResult = null;
            try {
                final String[] params = new String[] {Integer.toString(abilityConfig.getId())};
                queryResult = db.rawQuery(SQL_GET_ABILITY_PROPERTIES, params);
                for (queryResult.moveToFirst(); !queryResult.isAfterLast(); queryResult.moveToNext()) {
                    final String key = queryResult.getDataRow().getString(0);
                    final String value = queryResult.getDataRow().getString(1);
                    abilityProperties.put(key, value);
                }
            } catch (final SQLException sqlException) {
                //                Logger.error("Can't get properties of ability: " + abilityConfig, IllegalStateException);
            } finally {
                if (queryResult != null) {
                    queryResult.close();
                }
            }
            abilityConfig.setProperties(abilityProperties);
        }
    }

    private Ability createAbility(final AbilityConfig abilityConfig, final AbilityBuilderFactory abilityBuilderFactory) {
        final AbilityBuilder abilityBuilder = abilityBuilderFactory.getAbilityBuilder(abilityConfig.getClassname());
        return abilityBuilder.createAbility(abilityConfig);
    }

    @Override
    public Ability createAbility(final Ability ability) {
        insertAbilityTable(ability);
        insertAbilityPropertyTable(ability);
        return ability;
    }

    private void insertAbilityTable(final Ability ability) {
        final ContentValues abilityValues = getAbilityValues(ability);
        abilityValues.putNull(COLUMN_ID);
        synchronized (Database.DB_LOCK) {
            final long rowId = db.insertOrThrow(TABLE_ABILITY, null, abilityValues);
            if (rowId == -1) {
                throw new IllegalStateException("Can't insert ability in ability table");
            }
            ability.setId(db.queryId(SQL_GET_ID, rowId));
        }
    }

    private ContentValues getAbilityValues(final Ability ability) {
        final ContentValues contentValues = new ContentValues();
        contentValues.put(COLUMN_NAME, ability.getName());
        contentValues.put(COLUMN_DESCRIPTION, ability.getDescription());
        contentValues.put(COLUMN_ABILITY_TYPE_ID, ability.getAbilityType().ordinal());
        contentValues.put(COLUMN_CLASSNAME, ability.getClass().getSimpleName());
        return contentValues;
    }

    private void insertAbilityPropertyTable(final Ability ability) {
        if (ability instanceof ExtraFeatsAbility) {
            insertExtraFeatAbility(ability);
        } else if (ability instanceof ExtraSkillPointsAbility) {
            insertExtraSkillPointsAbility(ability);
        } else if (ability instanceof SpelllistAbility) {
            insertSpelllistAbility(ability);
        }
    }

    private void insertExtraFeatAbility(final Ability ability) {
        final ExtraFeatsAbility extraFeatsAbility = (ExtraFeatsAbility) ability;
        insertAbilityProperty(ability.getId(), KEY_NUMBER_OF_EXTRA_FEATS, extraFeatsAbility.getNumberOfFeats());
    }

    private void insertExtraSkillPointsAbility(final Ability ability) {
        final ExtraSkillPointsAbility extraSkillPointsAbility = (ExtraSkillPointsAbility) ability;
        insertAbilityProperty(ability.getId(), KEY_NUMBER_OF_SKILL_POINTS, extraSkillPointsAbility.getSkillPoints());
    }

    private void insertSpelllistAbility(final Ability ability) {
        final SpelllistAbility spelllistAbility = (SpelllistAbility) ability;
        insertAbilityProperty(ability.getId(), KEY_SPELLLIST_ID, spelllistAbility.getSpelllist().getId());
        insertAbilityProperty(ability.getId(), KEY_SPELL_ATTRIBUTE_ID, spelllistAbility.getSpellAttribute().ordinal());
        insertAbilityProperty(ability.getId(), KEY_CASTING_TYPE_ID, spelllistAbility.getCastingType().ordinal());
        insertAbilityProperty(ability.getId(), KEY_SPELL_SOURCE_ID, spelllistAbility.getSpellSource().ordinal());
        insertAbilityProperty(ability.getId(), KEY_KNOWN_SPELLS_TABLE_ID, spelllistAbility.getKnownSpellsTable().getId());
        insertAbilityProperty(ability.getId(), KEY_SPELLS_PER_DAY_TABLE_ID, spelllistAbility.getSpellsPerDayTable().getId());
        insertAbilityProperty(ability.getId(), spelllistAbility.isAttributeBonusSpellSlots());
        insertAbilityProperty(ability.getId(), KEY_SCHOOL_ID, getSchoolId(spelllistAbility.getSchools()));
    }

    private int getSchoolId(final EnumSet<School> schools) {
        if (schools.size() == 1) {
            final School[] school = new School[1];
            schools.toArray(school);
            return school[0].ordinal();
        }
        return ANY_SCHOOL_ID;
    }

    private void insertAbilityProperty(final int abilityId, final boolean property_value) {
        insertAbilityProperty(abilityId, SpelllistAbilityBuilder.KEY_ATTRIBUTE_BONUS_SPELL_SLOTS, Boolean.toString(property_value));
    }

    private void insertAbilityProperty(final int abilityId, final String property_key, final int property_value) {
        insertAbilityProperty(abilityId, property_key, Integer.toString(property_value));
    }

    private void insertAbilityProperty(final int abilityId, final String property_key, final String property_value) {
        final ContentValues contentValues = getAbilityPropertyContentValues(abilityId, property_key, property_value);

        synchronized (Database.DB_LOCK) {
            final long rowId = db.insertOrThrow(TABLE_ABILITY_PROPERTY, null, contentValues);
            if (rowId == -1) {
                throw new IllegalStateException("Can't insert ability property in ability_property table");
            }
        }
    }

    @Override
    public void updateAbility(final Ability ability) {
        updateAbilityTable(ability);
        updateAbilityPropertyTable(ability);
    }

    private void updateAbilityTable(final Ability ability) {
        final ContentValues abilityValues = getAbilityValues(ability);
        final String[] bindVariables = new String[] {Integer.toString(ability.getId())};
        synchronized (Database.DB_LOCK) {
            final int numberOfAffectedRows = db.update(TABLE_ABILITY, abilityValues, SQL_WHERE_ID, bindVariables);
            if (numberOfAffectedRows != 1) {
                throw new IllegalStateException("More or Less than 1 row affected: " + numberOfAffectedRows);
            }
        }
    }

    private void updateAbilityPropertyTable(final Ability ability) {
        if (ability instanceof ExtraFeatsAbility) {
            updateExtraFeatAbility(ability);
        } else if (ability instanceof ExtraSkillPointsAbility) {
            updateExtraSkillPointsAbility(ability);
        } else if (ability instanceof SpelllistAbility) {
            updateSpelllistAbility(ability);
        }
    }

    private void updateExtraFeatAbility(final Ability ability) {
        final ExtraFeatsAbility extraFeatsAbility = (ExtraFeatsAbility) ability;
        updateAbilityProperty(ability.getId(), KEY_NUMBER_OF_EXTRA_FEATS, extraFeatsAbility.getNumberOfFeats());
    }

    private void updateExtraSkillPointsAbility(final Ability ability) {
        final ExtraSkillPointsAbility extraSkillPointsAbility = (ExtraSkillPointsAbility) ability;
        updateAbilityProperty(ability.getId(), KEY_NUMBER_OF_SKILL_POINTS, extraSkillPointsAbility.getSkillPoints());
    }

    private void updateSpelllistAbility(final Ability ability) {
        final SpelllistAbility spelllistAbility = (SpelllistAbility) ability;
        updateAbilityProperty(ability.getId(), KEY_SPELLLIST_ID, spelllistAbility.getSpelllist().getId());
        updateAbilityProperty(ability.getId(), KEY_SPELL_ATTRIBUTE_ID, spelllistAbility.getSpellAttribute().ordinal());
        updateAbilityProperty(ability.getId(), KEY_CASTING_TYPE_ID, spelllistAbility.getCastingType().ordinal());
        updateAbilityProperty(ability.getId(), KEY_SPELL_SOURCE_ID, spelllistAbility.getSpellSource().ordinal());
        updateAbilityProperty(ability.getId(), KEY_KNOWN_SPELLS_TABLE_ID, spelllistAbility.getKnownSpellsTable().getId());
        updateAbilityProperty(ability.getId(), KEY_SPELLS_PER_DAY_TABLE_ID, spelllistAbility.getSpellsPerDayTable().getId());
        updateAbilityProperty(ability.getId(), spelllistAbility.isAttributeBonusSpellSlots());
        updateAbilityProperty(ability.getId(), KEY_SCHOOL_ID, getSchoolId(spelllistAbility.getSchools()));
    }

    private void updateAbilityProperty(final int abilityId, final boolean property_value) {
        updateAbilityProperty(abilityId, SpelllistAbilityBuilder.KEY_ATTRIBUTE_BONUS_SPELL_SLOTS, Boolean.toString(property_value));
    }

    private void updateAbilityProperty(final int abilityId, final String property_key, final int property_value) {
        updateAbilityProperty(abilityId, property_key, Integer.toString(property_value));
    }

    private void updateAbilityProperty(final int abilityId, final String property_key, final String property_value) {
        final ContentValues contentValues = getAbilityPropertyContentValues(abilityId, property_key, property_value);

        final String[] bindVariables = new String[] {Integer.toString(abilityId), property_key};
        synchronized (Database.DB_LOCK) {
            final int numberOfAffectedRows = db
                .update(TABLE_ABILITY_PROPERTY, contentValues, SQL_WHERE_ABILITY_ID_AND_PROPERTY_KEY, bindVariables);
            if (numberOfAffectedRows == 0) {
                db.insertOrThrow(TABLE_ABILITY_PROPERTY, null, contentValues);
            }
            if (numberOfAffectedRows > 1) {
                throw new IllegalStateException("More than 1 row affected: " + numberOfAffectedRows);
            }
        }
    }

    private ContentValues getAbilityPropertyContentValues(final int abilityId, final String property_key,
        final String property_value) {
        final ContentValues contentValues = new ContentValues();
        contentValues.put(COLUMN_ABILITY_ID, abilityId);
        contentValues.put(COLUMN_PROPERTY_KEY, property_key);
        contentValues.put(COLUMN_PROPERTY_VALUE, property_value);
        return contentValues;
    }

    @Override
    public void deleteAbility(final Ability ability) {
        final String[] abilityId = new String[] {Integer.toString(ability.getId())};
        synchronized (Database.DB_LOCK) {
            db.delete(TABLE_ABILITY, COLUMN_ID + " = ?", abilityId);
            db.delete(TABLE_ABILITY_PROPERTY, COLUMN_ABILITY_ID + " = ?", abilityId);
        }
    }

    @Override
    public AbilityBuilderFactory createAbilityBuilderFactory(final List<Spelllist> allSpelllists,
        final List<KnownSpellsTable> allKnownSpellsTables, final List<SpellsPerDayTable> allSpellsPerDayTables) {
        final AbilityBuilderFactory abilityBuilderFactory = new AbilityBuilderFactory();
        abilityBuilderFactory.registerAbilityBuilder(Ability.class.getSimpleName(), new DefaultAbilityBuilder());
        abilityBuilderFactory.registerAbilityBuilder(ExtraFeatsAbility.class.getSimpleName(), new ExtraFeatsAbilityBuilder());
        abilityBuilderFactory
            .registerAbilityBuilder(ExtraSkillPointsAbility.class.getSimpleName(), new ExtraSkillPointsAbilityBuilder());
        abilityBuilderFactory.registerAbilityBuilder(SpelllistAbility.class.getSimpleName(),
                                                     new SpelllistAbilityBuilder(allSpelllists, allKnownSpellsTables,
                                                                                 allSpellsPerDayTables));
        return abilityBuilderFactory;
    }
}

File: src/main/java/com/d20charactersheet/framework/dac/dao/sql/SqlAbilityDao.kt
package com.d20charactersheet.framework.dac.dao.sql

import com.d20charactersheet.framework.boc.model.*
import com.d20charactersheet.framework.dac.abilitybuilder.*
import com.d20charactersheet.framework.dac.dao.AbilityDao
import com.d20charactersheet.framework.dac.dao.sql.TableAndColumnNames.COLUMN_ABILITY_ID
import com.d20charactersheet.framework.dac.dao.sql.TableAndColumnNames.COLUMN_ABILITY_TYPE_ID
import com.d20charactersheet.framework.dac.dao.sql.TableAndColumnNames.COLUMN_CLASSNAME
import com.d20charactersheet.framework.dac.dao.sql.TableAndColumnNames.COLUMN_DESCRIPTION
import com.d20charactersheet.framework.dac.dao.sql.TableAndColumnNames.COLUMN_ID
import com.d20charactersheet.framework.dac.dao.sql.TableAndColumnNames.COLUMN_NAME
import com.d20charactersheet.framework.dac.dao.sql.TableAndColumnNames.COLUMN_PROPERTY_KEY
import com.d20charactersheet.framework.dac.dao.sql.TableAndColumnNames.COLUMN_PROPERTY_VALUE
import com.d20charactersheet.framework.dac.dao.sql.TableAndColumnNames.SELECT
import com.d20charactersheet.framework.dac.dao.sql.TableAndColumnNames.SQL_GET_ABILITY_PROPERTIES
import com.d20charactersheet.framework.dac.dao.sql.TableAndColumnNames.SQL_GET_ALL_ABILITIES
import com.d20charactersheet.framework.dac.dao.sql.TableAndColumnNames.SQL_WHERE_ID
import com.d20charactersheet.framework.dac.dao.sql.TableAndColumnNames.TABLE_ABILITY
import com.d20charactersheet.framework.dac.dao.sql.TableAndColumnNames.TABLE_ABILITY_PROPERTY
import com.d20charactersheet.framework.dac.dao.sql.rowmapper.AbilityConfigRowMapper
import java.sql.SQLException
import java.util.*

/**
 * Provides access to the ability tables in the SQLite database.
 */
class SqlAbilityDao(private val db: Database) : AbilityDao {

    private val abilityConfigRowMapper: RowMapper = AbilityConfigRowMapper()

    override fun getAllAbilities(allSpelllists: List<Spelllist>, allKnownSpellsTables: List<KnownSpellsTable>,
                                 allSpellsPerDayTables: List<SpellsPerDayTable>): List<Ability> {
        val allAbilities: MutableList<Ability> = ArrayList()
        val abilityBuilderFactoy = createAbilityBuilderFactory(allSpelllists, allKnownSpellsTables,
                allSpellsPerDayTables)
        var queryResult: QueryResult? = null
        try {
            queryResult = db.rawQuery(SQL_GET_ALL_ABILITIES, arrayOf())
            queryResult.moveToFirst()
            while (!queryResult.isAfterLast()) {
                val abilityConfig = abilityConfigRowMapper.mapRow(queryResult.getDataRow()) as AbilityConfig
                selectAbilityProperties(abilityConfig)
                val ability = createAbility(abilityConfig, abilityBuilderFactoy)
                allAbilities.add(ability)
                queryResult.moveToNext()
            }
        } catch (sqlException: SQLException) {
            //            Logger.error("Can't get all abilities", sqlException);
        } finally {
            queryResult?.close()
        }
        return allAbilities
    }

    private fun selectAbilityProperties(abilityConfig: AbilityConfig) {
        if (abilityConfig.classname != Ability::class.java.simpleName) {
            val abilityProperties: MutableMap<String, String> = HashMap()
            var queryResult: QueryResult? = null
            try {
                val params = arrayOf(abilityConfig.id.toString())
                queryResult = db.rawQuery(SQL_GET_ABILITY_PROPERTIES, params)
                queryResult.moveToFirst()
                while (!queryResult.isAfterLast()) {
                    val key = queryResult.getDataRow().getString(0)
                    val value = queryResult.getDataRow().getString(1)
                    abilityProperties[key] = value
                    queryResult.moveToNext()
                }
            } catch (sqlException: SQLException) {
                //                Logger.error("Can't get properties of ability: " + abilityConfig, IllegalStateException);
            } finally {
                queryResult?.close()
            }
            abilityConfig.properties = abilityProperties
        }
    }

    private fun createAbility(abilityConfig: AbilityConfig, abilityBuilderFactory: AbilityBuilderFactory): Ability {
        val abilityBuilder = abilityBuilderFactory.getAbilityBuilder(abilityConfig.classname)
        return abilityBuilder.createAbility(abilityConfig)
    }

    override fun createAbility(ability: Ability): Ability {
        insertAbilityTable(ability)
        insertAbilityPropertyTable(ability)
        return ability
    }

    private fun insertAbilityTable(ability: Ability) {
        val abilityValues = getAbilityValues(ability)
        abilityValues.putNull(COLUMN_ID)
        synchronized(Database.DB_LOCK) {
            val rowId = db.insertOrThrow(TABLE_ABILITY, abilityValues)
            check(rowId != -1L) { "Can't insert ability in ability table" }
            ability.id = db.queryId(SQL_GET_ID, rowId)
        }
    }

    private fun getAbilityValues(ability: Ability): ContentValues {
        val contentValues = ContentValues()
        contentValues.put(COLUMN_NAME, ability.name)
        contentValues.put(COLUMN_DESCRIPTION, ability.description)
        contentValues.put(COLUMN_ABILITY_TYPE_ID, ability.abilityType.ordinal)
        contentValues.put(COLUMN_CLASSNAME, ability.javaClass.simpleName)
        return contentValues
    }

    private fun insertAbilityPropertyTable(ability: Ability) {
        (ability as? ExtraFeatsAbility)?.let { insertExtraFeatAbility(it) }
                ?: ((ability as? ExtraSkillPointsAbility)?.let { insertExtraSkillPointsAbility(it) }
                        ?: (ability as? SpelllistAbility)?.let { insertSpelllistAbility(it) })
    }

    private fun insertExtraFeatAbility(ability: Ability) {
        val extraFeatsAbility = ability as ExtraFeatsAbility
        insertAbilityProperty(ability.getId(), ExtraFeatsAbilityBuilder.KEY_NUMBER_OF_EXTRA_FEATS, extraFeatsAbility.numberOfFeats)
    }

    private fun insertExtraSkillPointsAbility(ability: Ability) {
        val extraSkillPointsAbility = ability as ExtraSkillPointsAbility
        insertAbilityProperty(ability.getId(), ExtraSkillPointsAbilityBuilder.KEY_NUMBER_OF_SKILL_POINTS, extraSkillPointsAbility.skillPoints)
    }

    private fun insertSpelllistAbility(ability: Ability) {
        val spelllistAbility = ability as SpelllistAbility
        insertAbilityProperty(ability.getId(), SpelllistAbilityBuilder.KEY_SPELLLIST_ID, spelllistAbility.spelllist.id)
        insertAbilityProperty(ability.getId(), SpelllistAbilityBuilder.KEY_SPELL_ATTRIBUTE_ID, spelllistAbility.spellAttribute.ordinal)
        insertAbilityProperty(ability.getId(), SpelllistAbilityBuilder.KEY_CASTING_TYPE_ID, spelllistAbility.castingType.ordinal)
        insertAbilityProperty(ability.getId(), SpelllistAbilityBuilder.KEY_SPELL_SOURCE_ID, spelllistAbility.spellSource.ordinal)
        insertAbilityProperty(ability.getId(), SpelllistAbilityBuilder.KEY_KNOWN_SPELLS_TABLE_ID, spelllistAbility.knownSpellsTable.id)
        insertAbilityProperty(ability.getId(), SpelllistAbilityBuilder.KEY_SPELLS_PER_DAY_TABLE_ID, spelllistAbility.spellsPerDayTable.id)
        insertAbilityProperty(ability.getId(), spelllistAbility.isAttributeBonusSpellSlots)
        insertAbilityProperty(ability.getId(), SpelllistAbilityBuilder.KEY_SCHOOL_ID, getSchoolId(spelllistAbility.schools))
    }

    private fun getSchoolId(schools: EnumSet<School?>): Int {
        if (schools.size == 1) {
            val school = arrayOfNulls<School>(1)
            schools.toArray(school)
            return school[0]!!.ordinal
        }
        return SpelllistAbilityBuilder.ANY_SCHOOL_ID
    }

    private fun insertAbilityProperty(abilityId: Int, property_value: Boolean) {
        insertAbilityProperty(abilityId, SpelllistAbilityBuilder.KEY_ATTRIBUTE_BONUS_SPELL_SLOTS, java.lang.Boolean.toString(property_value))
    }

    private fun insertAbilityProperty(abilityId: Int, property_key: String, property_value: Int) {
        insertAbilityProperty(abilityId, property_key, property_value.toString())
    }

    private fun insertAbilityProperty(abilityId: Int, property_key: String, property_value: String) {
        val contentValues = getAbilityPropertyContentValues(abilityId, property_key, property_value)
        synchronized(Database.DB_LOCK) {
            val rowId = db.insertOrThrow(TABLE_ABILITY_PROPERTY, contentValues)
            check(rowId != -1L) { "Can't insert ability property in ability_property table" }
        }
    }

    override fun updateAbility(ability: Ability) {
        updateAbilityTable(ability)
        updateAbilityPropertyTable(ability)
    }

    private fun updateAbilityTable(ability: Ability) {
        val abilityValues = getAbilityValues(ability)
        val bindVariables = arrayOf(ability.id.toString())
        synchronized(Database.DB_LOCK) {
            val numberOfAffectedRows = db.update(TABLE_ABILITY, abilityValues, SQL_WHERE_ID, bindVariables)
            check(numberOfAffectedRows == 1) { "More or Less than 1 row affected: $numberOfAffectedRows" }
        }
    }

    private fun updateAbilityPropertyTable(ability: Ability) {
        (ability as? ExtraFeatsAbility)?.let { updateExtraFeatAbility(it) }
                ?: ((ability as? ExtraSkillPointsAbility)?.let { updateExtraSkillPointsAbility(it) }
                        ?: (ability as? SpelllistAbility)?.let { updateSpelllistAbility(it) })
    }

    private fun updateExtraFeatAbility(ability: Ability) {
        val extraFeatsAbility = ability as ExtraFeatsAbility
        updateAbilityProperty(ability.getId(), ExtraFeatsAbilityBuilder.KEY_NUMBER_OF_EXTRA_FEATS, extraFeatsAbility.numberOfFeats)
    }

    private fun updateExtraSkillPointsAbility(ability: Ability) {
        val extraSkillPointsAbility = ability as ExtraSkillPointsAbility
        updateAbilityProperty(ability.getId(), ExtraSkillPointsAbilityBuilder.KEY_NUMBER_OF_SKILL_POINTS, extraSkillPointsAbility.skillPoints)
    }

    private fun updateSpelllistAbility(ability: Ability) {
        val spelllistAbility = ability as SpelllistAbility
        updateAbilityProperty(ability.getId(), SpelllistAbilityBuilder.KEY_SPELLLIST_ID, spelllistAbility.spelllist.id)
        updateAbilityProperty(ability.getId(), SpelllistAbilityBuilder.KEY_SPELL_ATTRIBUTE_ID, spelllistAbility.spellAttribute.ordinal)
        updateAbilityProperty(ability.getId(), SpelllistAbilityBuilder.KEY_CASTING_TYPE_ID, spelllistAbility.castingType.ordinal)
        updateAbilityProperty(ability.getId(), SpelllistAbilityBuilder.KEY_SPELL_SOURCE_ID, spelllistAbility.spellSource.ordinal)
        updateAbilityProperty(ability.getId(), SpelllistAbilityBuilder.KEY_KNOWN_SPELLS_TABLE_ID, spelllistAbility.knownSpellsTable.id)
        updateAbilityProperty(ability.getId(), SpelllistAbilityBuilder.KEY_SPELLS_PER_DAY_TABLE_ID, spelllistAbility.spellsPerDayTable.id)
        updateAbilityProperty(ability.getId(), spelllistAbility.isAttributeBonusSpellSlots)
        updateAbilityProperty(ability.getId(), SpelllistAbilityBuilder.KEY_SCHOOL_ID, getSchoolId(spelllistAbility.schools))
    }

    private fun updateAbilityProperty(abilityId: Int, property_value: Boolean) {
        updateAbilityProperty(abilityId, SpelllistAbilityBuilder.KEY_ATTRIBUTE_BONUS_SPELL_SLOTS, java.lang.Boolean.toString(property_value))
    }

    private fun updateAbilityProperty(abilityId: Int, property_key: String, property_value: Int) {
        updateAbilityProperty(abilityId, property_key, property_value.toString())
    }

    private fun updateAbilityProperty(abilityId: Int, property_key: String, property_value: String) {
        val contentValues = getAbilityPropertyContentValues(abilityId, property_key, property_value)
        val bindVariables = arrayOf(abilityId.toString(), property_key)
        synchronized(Database.DB_LOCK) {
            val numberOfAffectedRows = db
                    .update(TABLE_ABILITY_PROPERTY, contentValues, SQL_WHERE_ABILITY_ID_AND_PROPERTY_KEY, bindVariables)
            if (numberOfAffectedRows == 0) {
                db.insertOrThrow(TABLE_ABILITY_PROPERTY, contentValues)
            }
            check(numberOfAffectedRows <= 1) { "More than 1 row affected: $numberOfAffectedRows" }
        }
    }

    private fun getAbilityPropertyContentValues(abilityId: Int, property_key: String,
                                                property_value: String): ContentValues {
        val contentValues = ContentValues()
        contentValues.put(COLUMN_ABILITY_ID, abilityId)
        contentValues.put(COLUMN_PROPERTY_KEY, property_key)
        contentValues.put(COLUMN_PROPERTY_VALUE, property_value)
        return contentValues
    }

    override fun deleteAbility(ability: Ability) {
        val abilityId = arrayOf(ability.id.toString())
        synchronized(Database.DB_LOCK) {
            db.delete(TABLE_ABILITY, "$COLUMN_ID = ?", abilityId)
            db.delete(TABLE_ABILITY_PROPERTY, "$COLUMN_ABILITY_ID = ?", abilityId)
        }
    }

    override fun createAbilityBuilderFactory(allSpelllists: List<Spelllist>,
                                             allKnownSpellsTables: List<KnownSpellsTable>, allSpellsPerDayTables: List<SpellsPerDayTable>): AbilityBuilderFactory {
        val abilityBuilderFactory = AbilityBuilderFactory()
        abilityBuilderFactory.registerAbilityBuilder(Ability::class.java.simpleName, DefaultAbilityBuilder())
        abilityBuilderFactory.registerAbilityBuilder(ExtraFeatsAbility::class.java.simpleName, ExtraFeatsAbilityBuilder())
        abilityBuilderFactory
                .registerAbilityBuilder(ExtraSkillPointsAbility::class.java.simpleName, ExtraSkillPointsAbilityBuilder())
        abilityBuilderFactory.registerAbilityBuilder(SpelllistAbility::class.java.simpleName,
                SpelllistAbilityBuilder(allSpelllists, allKnownSpellsTables,
                        allSpellsPerDayTables))
        return abilityBuilderFactory
    }

    companion object {
        private const val SQL_GET_ID: String = SELECT + "id FROM " + TABLE_ABILITY + " WHERE rowid = ?"
        private const val SQL_WHERE_ABILITY_ID_AND_PROPERTY_KEY: String = "$COLUMN_ABILITY_ID = ? AND $COLUMN_PROPERTY_KEY = ?"
    }

}